src/webgpu/api/operation/command_buffer/render/culling_tests.spec.ts

```diff
@@ -0,0 +1,160 @@
+      vertexStage: { module: vertexModule, entryPoint: 'main' },
+      fragmentStage: { module: fragmentModule, entryPoint: 'main' },
+      primitiveTopology: 'triangle-list',
+      rasterizationState: { frontFace, cullMode },
+      colorStates: [
+        {
+          format,
+          alphaBlend: { dstFactor: 'zero', srcFactor: 'one', operation: 'add' },
+          colorBlend: { dstFactor: 'zero', srcFactor: 'one', operation: 'add' },
+          writeMask: GPUColorWrite.ALL,
+        },
+      ],
+      sampleCount: 1,
+    });
+  }
+
+  create2DTextureForTest(format: GPUTextureFormat, size: number) {
+    return this.device.createTexture({
+      size: { width: size, height: size, depth: 1 },
+      format,
+      mipLevelCount: 1,
+      sampleCount: 1,
+      dimension: '2d',
+      usage: GPUTextureUsage.OUTPUT_ATTACHMENT | GPUTextureUsage.COPY_SRC,
+    });
+  }
+
+  run(
+    frontFace: GPUFrontFace,
+    cullMode: GPUCullMode,
+    isCCWTriangleCulled: boolean,
+    isCWTriangleCulled: boolean,
+    format: GPUTextureFormat
+  ): void {
+    const size = 4;
+    const texture = this.create2DTextureForTest(format, size);
+    const encoder = this.device.createCommandEncoder();
+    const pass = encoder.beginRenderPass({
+      colorAttachments: [
+        {
+          attachment: texture.createView(),
+          loadValue: { r: 0.0, g: 0.0, b: 1.0, a: 1.0 },
+          storeOp: 'store',
+        },
+      ],
+    });
+    pass.setPipeline(this.createPipelineForTest(frontFace, cullMode, format));
+    pass.draw(6, 1, 0, 0);
+    pass.endPass();
+
+    const topLeftBuffer = this.device.createBuffer({
+      size,
+      usage: GPUBufferUsage.COPY_SRC | GPUBufferUsage.COPY_DST,
+    });
+    encoder.copyTextureToBuffer(
+      { texture, mipLevel: 0, origin: { x: 0, y: 0, z: 0 } },
+      { buffer: topLeftBuffer, bytesPerRow: 256 },
+      { width: 1, height: 1, depth: 1 }
+    );
+
+    const bottomRightBuffer = this.device.createBuffer({
+      size,
+      usage: GPUBufferUsage.COPY_SRC | GPUBufferUsage.COPY_DST,
+    });
+    encoder.copyTextureToBuffer(
+      { texture, mipLevel: 0, origin: { x: size - 1, y: size - 1, z: 0 } },
+      { buffer: bottomRightBuffer, bytesPerRow: 256 },
+      { width: 1, height: 1, depth: 1 }
+    );
+
+    this.device.defaultQueue.submit([encoder.finish()]);
+
+    const kCCWTriangleTopLeftColor = isCCWTriangleCulled
+      ? new Uint8Array([0x00, 0x00, 0xff, 0xff])
+      : new Uint8Array([0x00, 0x00, 0x00, 0xff]);
+    this.expectContents(topLeftBuffer, kCCWTriangleTopLeftColor);
+
+    const kCWTriangleBottomRightColor = isCWTriangleCulled
+      ? new Uint8Array([0x00, 0x00, 0xff, 0xff])
+      : new Uint8Array([0x03, 0x03, 0x00, 0xff]);
+    this.expectContents(bottomRightBuffer, kCWTriangleBottomRightColor);
+  }
+}
+
+export const g = makeTestGroup(CullingTest);
+
+g.test('CullNoneWhenCCWIsFrontFace')
+  .params([{ format: 'rgba8unorm' as GPUTextureFormat }])
```

Comment:
I think most  different formats are unlikely to have much of an effect in these tests. We should leave it out if we only test one variation, though we should probably have some more combinatorics here - or at least TODOs to add them. Here are some ideas (that would go into a test plan if we had one):

test culling is performed and correctly written into all attachments, with following parameters:
- x= actual face orientation [cw, ccw] (we have this above)
- x= all possible frontFace (have this below)
- x= all possible cullMode (have this below)
- x= DS attachment = [none, depth, stencil, depth+stencil]
- x= color attachments = [0, 1, >1]

I haven't thought of other combinatorics that are likely to interact with culling, but there might be more. We want to be as thorough as is reasonable.

Reply:
It would be nice to test with triangle list vs. triangle strip too because there's some complexity in determining the front face of the triangles in a strip. We should also have a similar test (or even the same one) for the gl_FrontFace builtin.